File: crates/certimining-log/src/tree.rs
//! The fixed-capacity epoch tree (§1.4, §2.3).
//!
//! Every epoch is a complete tree of the same height. Real leaves sit in slots chosen by a PRF under
//! the epoch key, every remaining slot carries PRF padding, and the root is published whatever the
//! record count was: an epoch holding one filing and an epoch holding two hundred are identical in
//! leaf count, proof length and root distribution (INV-TREE-01, INV-TREE-02).
//!
//! What this does not carry: the batcher's queue and the overflow discipline of INV-TREE-04, which
//! are E-07's. `build` refuses more submissions than `C` with `0x12` and leaves the queueing to its
//! caller.

use alloc::vec;
use alloc::vec::Vec;

use certimining_core::{
    epoch_key, padding_prf, slot_seed, Digest, Hasher, NodePreimage, PaddingPreimage, Preimage,
    RealLeafPreimage, RegistryError, Result, SubmissionId,
};

use crate::{InclusionProof, MAX_SIBLINGS};

/// §1.8's range for the tree height, fixed at `initialize` and immutable thereafter (INV-TREE-06).
pub const MIN_HEIGHT: u8 = 4;
/// The upper end of §1.8's range.
pub const MAX_HEIGHT: u8 = 16;

/// What one sealed epoch carries (§2.3, D-62).
///
/// Nothing here marks a slot real or padding. That distinction is what V-Z-02 tries to recover from
/// a root and a leaf set, so it is not written down where a later caller could reach it.
#[derive(Debug, Clone)]
pub struct BuiltEpoch {
    /// The epoch this tree is for, a UTC day index (§1.4).
    pub epoch: u64,
    /// The height every epoch of this log uses (INV-TREE-06).
    pub height: u8,
    /// The root the checkpoint publishes.
    pub root: Digest,
    /// Every slot's leaf digest, in slot order, exactly `C` of them.
    pub leaves: Vec<Digest>,
    /// Submission identifier to slot, ascending by identifier (D-60).
    pub assignment: Vec<(SubmissionId, u16)>,
    /// Levels 1 to `H` concatenated, `C - 1` digests, so a proof costs `H` lookups rather than a
    /// rebuild. An implementation detail of the builder rather than part of the contract (D-62).
    nodes: Vec<Digest>,
}

/// §1.4's tree, built once per epoch (§2.3).
pub trait EpochTree {
    /// The height this tree was built at.
    fn height(&self) -> u8;
    /// `C`, which is `1 << height()`.
    fn capacity(&self) -> usize;
    /// Builds one epoch. `key` is `k_master`: the epoch key is derived here from the epoch, so no
    /// caller can reuse one `k_e` across epochs (INV-TREE-05, D-63). The hasher is named at the call
    /// site, as it is for every digest in §2.2.
    fn build<H: Hasher>(
        epoch: u64,
        height: u8,
        key: &Digest,
        real: &[(SubmissionId, Digest)],
    ) -> Result<BuiltEpoch>;
    /// The epoch root.
    fn root(&self) -> Digest;
    /// The inclusion proof for a submission this epoch holds. A submission it does not hold is
    /// `0x16`, which is its own condition and so takes its own code (D-67).
    fn proof(&self, id: &SubmissionId) -> Result<InclusionProof>;
}

impl EpochTree for BuiltEpoch {
    fn height(&self) -> u8 {
        self.height
    }

    fn capacity(&self) -> usize {
        self.leaves.len()
    }

    fn build<H: Hasher>(
        epoch: u64,
        height: u8,
        key: &Digest,
        real: &[(SubmissionId, Digest)],
    ) -> Result<BuiltEpoch> {
        let capacity = capacity_of(height)?;
        if real.len() > capacity {
            return Err(RegistryError::EpochCapacityExceeded);
        }

        // The epoch key, derived from the master key and this epoch (INV-TREE-05).
        let k_e = epoch_key::<H>(key, epoch)?;

        // Ascending by identifier, so one set of submissions gives one tree whatever order the
        // caller supplied them in (D-60). Two submissions under one identifier make the set
        // malformed: a proof could be answered for neither.
        let mut ordered: Vec<(SubmissionId, Digest)> = real.to_vec();
        ordered.sort_unstable_by_key(|entry| entry.0);
        if ordered
            .windows(2)
            .any(|pair| matches!(pair, [left, right] if left.0 == right.0))
        {
            return Err(RegistryError::MalformedPayload);
        }

        let mut slots: Vec<Option<Digest>> = vec![None; capacity];
        let mut assignment: Vec<(SubmissionId, u16)> = Vec::new();
        for (id, leaf) in &ordered {
            let seed = slot_seed::<H>(&k_e, id)?;
            let slot = probe(&slots, slot_from_seed(&seed, height)?, capacity)?;
            let digest = RealLeafPreimage { leaf: *leaf }.digest::<H>()?;
            let cell = slots
                .get_mut(usize::from(slot))
                .ok_or(RegistryError::ArithmeticOverflow)?;
            *cell = Some(digest);
            assignment.push((*id, slot));
        }

        // Padding fills every slot the assignment left empty, each committing to the PRF output for
        // its own index (§1.4).
        let mut leaves: Vec<Digest> = Vec::new();
        for (index, slot) in slots.iter().enumerate() {
            match slot {
                Some(digest) => leaves.push(*digest),
                None => {
                    let index =
                        u16::try_from(index).map_err(|_| RegistryError::ArithmeticOverflow)?;
                    let prf_output = padding_prf::<H>(&k_e, index)?;
                    leaves.push(PaddingPreimage { prf_output }.digest::<H>()?);
                }
            }
        }

        let (nodes, root) = levels_above::<H>(&leaves)?;
        Ok(BuiltEpoch {
            epoch,
            height,
            root,
            leaves,
            assignment,
            nodes,
        })
    }

    fn root(&self) -> Digest {
        self.root
    }

    fn proof(&self, id: &SubmissionId) -> Result<InclusionProof> {
        let slot = self
            .assignment
            .iter()
            .find(|(candidate, _)| candidate == id)
            .map(|(_, slot)| *slot)
            .ok_or(RegistryError::SubmissionNotInEpoch)?;
        let capacity = self.leaves.len();
        let mut siblings: heapless::Vec<Digest, MAX_SIBLINGS> = heapless::Vec::new();
        let mut index = usize::from(slot);
        for level in 0..usize::from(self.height) {
            // The sibling shares every bit above this level and differs in this one (D-64).
            let sibling = index ^ 1;
            let digest = if level == 0 {
                self.leaves.get(sibling)
            } else {
                let offset = level_offset(level, capacity)?;
                self.nodes.get(
                    offset
                        .checked_add(sibling)
                        .ok_or(RegistryError::ArithmeticOverflow)?,
                )
            };
            siblings
                .push(*digest.ok_or(RegistryError::InclusionProofInvalid)?)
                .map_err(|_| RegistryError::InclusionProofInvalid)?;
            index = index
                .checked_shr(1)
                .ok_or(RegistryError::ArithmeticOverflow)?;
        }
        Ok(InclusionProof {
            height: self.height,
            siblings,
            slot_index: slot,
            epoch: self.epoch,
        })
    }
}

/// §1.4's reduction (D-60): the seed read as a little-endian integer and reduced modulo `C`. Because
/// `C` is a power of two this is the low `H` bits, which lie in the digest's first two bytes.
pub fn slot_from_seed(seed: &Digest, height: u8) -> Result<u16> {
    if !(MIN_HEIGHT..=MAX_HEIGHT).contains(&height) {
        return Err(RegistryError::MalformedPayload);
    }
    let low: [u8; 2] = seed
        .get(..2)
        .and_then(|bytes| bytes.try_into().ok())
        .ok_or(RegistryError::MalformedPayload)?;
    let value = u16::from_le_bytes(low);
    if height == MAX_HEIGHT {
        // Every bit of the pair is in range, and `1 << 16` does not fit a `u16`.
        return Ok(value);
    }
    let mask = 1u16
        .checked_shl(u32::from(height))
        .and_then(|bound| bound.checked_sub(1))
        .ok_or(RegistryError::ArithmeticOverflow)?;
    Ok(value & mask)
}

/// `C` for a height inside §1.8's range.
fn capacity_of(height: u8) -> Result<usize> {
    if !(MIN_HEIGHT..=MAX_HEIGHT).contains(&height) {
        return Err(RegistryError::MalformedPayload);
    }
    1usize
        .checked_shl(u32::from(height))
        .ok_or(RegistryError::ArithmeticOverflow)
}

/// The first free slot at or after `start`, stepping upward by one and wrapping at `C` (D-60).
fn probe(slots: &[Option<Digest>], start: u16, capacity: usize) -> Result<u16> {
    let start = usize::from(start);
    for step in 0..capacity {
        let index = start
            .checked_add(step)
            .and_then(|sum| sum.checked_rem(capacity))
            .ok_or(RegistryError::ArithmeticOverflow)?;
        if slots
            .get(index)
            .ok_or(RegistryError::ArithmeticOverflow)?
            .is_none()
        {
            return u16::try_from(index).map_err(|_| RegistryError::ArithmeticOverflow);
        }
    }
    // Unreachable while the caller has refused a set larger than `C`, and the right answer if it
    // ever does not.
    Err(RegistryError::EpochCapacityExceeded)
}

/// Every level above the leaves, concatenated, and the root (§1.4).
fn levels_above<H: Hasher>(leaves: &[Digest]) -> Result<(Vec<Digest>, Digest)> {
    let mut nodes: Vec<Digest> = Vec::new();
    let mut current: Vec<Digest> = leaves.to_vec();
    while current.len() > 1 {
        let mut parents: Vec<Digest> = Vec::new();
        for pair in current.chunks(2) {
            match pair {
                [left, right] => parents.push(
                    NodePreimage {
                        left: *left,
                        right: *right,
                    }
                    .digest::<H>()?,
                ),
                // A complete tree over a power-of-two leaf count always pairs.
                _ => return Err(RegistryError::MalformedPayload),
            }
        }
        nodes.extend_from_slice(&parents);
        current = parents;
    }
    let root = *current.first().ok_or(RegistryError::MalformedPayload)?;
    Ok((nodes, root))
}

/// Where level `level` begins inside `nodes`. Level 1 holds `C / 2` digests, level 2 holds `C / 4`,
/// and everything below level `L` sums to `C − C / 2^(L−1)`.
fn level_offset(level: usize, capacity: usize) -> Result<usize> {
    let shift = level
        .checked_sub(1)
        .and_then(|below| u32::try_from(below).ok())
        .ok_or(RegistryError::ArithmeticOverflow)?;
    let remaining = capacity
        .checked_shr(shift)
        .ok_or(RegistryError::ArithmeticOverflow)?;
    capacity
        .checked_sub(remaining)
        .ok_or(RegistryError::ArithmeticOverflow)
}

File: ts/src/tree.ts
/**
 * §1.4's fixed-capacity, count-hiding epoch tree: the PRF uses, slot assignment with linear
 * probing, padding leaves, the complete-tree build, and the pure inclusion verifier of §2.3.
 */
import { type Digest, bytesEqual, compareBytes, concat, readU16le, u16le, u64le } from "./bytes.ts";
import { RegistryFailure } from "./errors.ts";
import { keccak256 } from "./hash.ts";
import { nodePreimage, paddingLeafPreimage, prfPreimage, realLeafPreimage } from "./preimage.ts";

/** §1.1's three PRF use codes, which separate the three uses inside one construction. */
export const USE_EPOCH_KEY = 0x01;
export const USE_SLOT = 0x02;
export const USE_PADDING = 0x03;

export const MIN_HEIGHT = 4;
export const MAX_HEIGHT = 16;
/** §1.8's deployed default. Nothing in the verifier assumes it; it is checked against the log. */
export const DEFAULT_HEIGHT = 8;

export type SubmissionId = Uint8Array; // 16 bytes

export type InclusionProof = {
  height: number;
  siblings: Digest[];
  slotIndex: number;
  epoch: bigint;
};

export type BuiltEpoch = {
  epoch: bigint;
  height: number;
  root: Digest;
  leaves: Digest[];
  assignment: Array<{ submissionId: SubmissionId; slot: number }>;
};

export function capacityOf(height: number): number {
  return 1 << height;
}

function assertHeight(height: number): void {
  if (!Number.isInteger(height) || height < MIN_HEIGHT || height > MAX_HEIGHT) {
    throw new RegistryFailure("MalformedPayload", `tree height ${height} is outside [4, 16]`);
  }
}

/** `PRF(k, x) = Keccak256(TAG_PRF ‖ k ‖ len(x) ‖ x)`. */
export function prf(k: Digest, x: Uint8Array): Digest {
  return keccak256(prfPreimage(k, x));
}

/** INV-TREE-05: `k_e = PRF(k_master, 0x01 ‖ e_le)`. Derived per epoch, never published. */
export function epochKey(masterKey: Digest, epoch: bigint): Digest {
  return prf(masterKey, concat([Uint8Array.of(USE_EPOCH_KEY), u64le(epoch)]));
}

/** §1.4: the slot seed, `PRF(k_e, 0x02 ‖ submission_id)`. */
export function slotSeed(epochKeyDigest: Digest, submissionId: SubmissionId): Digest {
  if (submissionId.length !== 16) {
    throw new RegistryFailure("MalformedPayload", `submission_id must be 16 bytes, got ${submissionId.length}`);
  }
  return prf(epochKeyDigest, concat([Uint8Array.of(USE_SLOT), submissionId]));
}

/**
 * D-60: the PRF output is read as a little-endian integer and reduced modulo `C`. `C` is a power
 * of two, so that is the low `H` bits, which lie in the digest's first two bytes.
 */
export function slotFromSeed(seed: Digest, height: number): number {
  assertHeight(height);
  return readU16le(seed, 0) & (capacityOf(height) - 1);
}

/** §1.4: `padding leaf = Keccak256(TAG_PAD ‖ PRF(k_e, 0x03 ‖ slot_index_le))`. */
export function paddingLeaf(epochKeyDigest: Digest, slotIndex: number): Digest {
  const output = prf(epochKeyDigest, concat([Uint8Array.of(USE_PADDING), u16le(slotIndex)]));
  return keccak256(paddingLeafPreimage(output));
}

/** §1.4: `real leaf = Keccak256(TAG_MTL0 ‖ leafₙ)`. */
export function realLeaf(chainLeaf: Digest): Digest {
  return keccak256(realLeafPreimage(chainLeaf));
}

/** §1.4: `internal node = Keccak256(TAG_MTN1 ‖ left ‖ right)`. */
export function internalNode(left: Digest, right: Digest): Digest {
  return keccak256(nodePreimage(left, right));
}

/**
 * §1.4's slot assignment. Real submissions are assigned in ascending order of submission
 * identifier, so one set of submissions produces one tree whatever order the caller supplies them
 * in; the probe steps upward by one slot and wraps at `C`, taking the first free slot it meets.
 */
export function assignSlots(
  epochKeyDigest: Digest,
  height: number,
  real: Array<{ submissionId: SubmissionId; leaf: Digest }>,
): Array<{ submissionId: SubmissionId; slot: number; leaf: Digest }> {
  assertHeight(height);
  const capacity = capacityOf(height);
  if (real.length > capacity) {
    throw new RegistryFailure("EpochCapacityExceeded", `${real.length} real submissions into ${capacity} slots`);
  }
  const ordered = [...real].sort((a, b) => compareBytes(a.submissionId, b.submissionId));
  for (let i = 1; i < ordered.length; i++) {
    if (compareBytes(ordered[i - 1]!.submissionId, ordered[i]!.submissionId) === 0) {
      // Two submissions carrying the same identifier in one epoch is 0x05: the set is malformed,
      // and a proof could be answered for neither of them.
      throw new RegistryFailure("MalformedPayload", "two submissions carry the same identifier in one epoch");
    }
  }
  const taken = new Array<boolean>(capacity).fill(false);
  const out: Array<{ submissionId: SubmissionId; slot: number; leaf: Digest }> = [];
  for (const entry of ordered) {
    const start = slotFromSeed(slotSeed(epochKeyDigest, entry.submissionId), height);
    let slot = start;
    let steps = 0;
    while (taken[slot]) {
      slot = (slot + 1) % capacity;
      steps += 1;
      if (steps > capacity) throw new RegistryFailure("EpochCapacityExceeded", "no free slot");
    }
    taken[slot] = true;
    out.push({ submissionId: entry.submissionId, slot, leaf: entry.leaf });
  }
  return out;
}

/** The complete binary tree over all `C` slots. Its shape never varies with the record count. */
export function rootOf(leaves: Digest[]): Digest {
  let level = leaves;
  while (level.length > 1) {
    const next: Digest[] = [];
    for (let i = 0; i < level.length; i += 2) next.push(internalNode(level[i]!, level[i + 1]!));
    level = next;
  }
  return level[0]!;
}

/**
 * §2.3 `EpochTree::build`. `key` is `k_master`: the epoch key is derived inside, from the epoch
 * the tree already knows, so no caller can reuse one `k_e` across epochs (INV-TREE-05, D-63).
 */
export function buildEpoch(
  epoch: bigint,
  height: number,
  masterKey: Digest,
  real: Array<{ submissionId: SubmissionId; leaf: Digest }>,
): BuiltEpoch {
  assertHeight(height);
  const ke = epochKey(masterKey, epoch);
  const assigned = assignSlots(ke, height, real);
  const capacity = capacityOf(height);
  const leaves: Digest[] = new Array(capacity);
  for (const a of assigned) leaves[a.slot] = realLeaf(a.leaf);
  for (let i = 0; i < capacity; i++) if (leaves[i] === undefined) leaves[i] = paddingLeaf(ke, i);
  return {
    epoch,
    height,
    root: rootOf(leaves),
    leaves,
    assignment: assigned
      .map((a) => ({ submissionId: a.submissionId, slot: a.slot }))
      .sort((a, b) => compareBytes(a.submissionId, b.submissionId)),
  };
}

/** §2.3 `EpochTree::proof`. A submission this epoch does not hold is 0x16, not 0x13 (D-67). */
export function proofFor(built: BuiltEpoch, submissionId: SubmissionId): InclusionProof {
  const entry = built.assignment.find((a) => bytesEqual(a.submissionId, submissionId));
  if (entry === undefined) {
    throw new RegistryFailure("SubmissionNotInEpoch", "this epoch holds no such submission");
  }
  const siblings: Digest[] = [];
  let index = entry.slot;
  let level = built.leaves;
  while (level.length > 1) {
    const siblingIndex = index % 2 === 0 ? index + 1 : index - 1;
    siblings.push(level[siblingIndex]!);
    const next: Digest[] = [];
    for (let i = 0; i < level.length; i += 2) next.push(internalNode(level[i]!, level[i + 1]!));
    level = next;
    index = Math.floor(index / 2);
  }
  return { height: built.height, siblings, slotIndex: entry.slot, epoch: built.epoch };
}

/**
 * §2.3 `InclusionVerifier::verify`. Pure: no network, no clock, no storage, no batcher handle
 * (INV-IFACE-01). `leaf` is the chain leaf `leafₙ`; the verifier applies `TAG_MTL0` itself, so no
 * caller can omit the tag (INV-ENC-01, D-64).
 */
export function verifyInclusion(leaf: Digest, proof: InclusionProof, root: Digest): void {
  if (!Number.isInteger(proof.height) || proof.height < MIN_HEIGHT || proof.height > MAX_HEIGHT) {
    throw new RegistryFailure("InclusionProofInvalid", `proof height ${proof.height} is outside §1.8's [4, 16]`);
  }
  if (proof.siblings.length !== proof.height) {
    throw new RegistryFailure("InclusionProofInvalid", `${proof.siblings.length} siblings for height ${proof.height}`);
  }
  if (!Number.isInteger(proof.slotIndex) || proof.slotIndex < 0 || proof.slotIndex >= capacityOf(proof.height)) {
    throw new RegistryFailure("InclusionProofInvalid", `slot_index ${proof.slotIndex} is outside the tree`);
  }
  let node = realLeaf(leaf);
  for (let level = 0; level < proof.height; level++) {
    const sibling = proof.siblings[level]!;
    if (sibling.length !== 32) {
      throw new RegistryFailure("InclusionProofInvalid", `sibling at level ${level} is not 32 bytes`);
    }
    node = ((proof.slotIndex >> level) & 1) === 0 ? internalNode(node, sibling) : internalNode(sibling, node);
  }
  if (!bytesEqual(node, root)) {
    throw new RegistryFailure("InclusionProofInvalid", "the path does not reach the root");
  }
}

/**
 * §2.3 `InclusionVerifier::verify_for_height`: the same check for a caller that knows its log's
 * configured height. A proof whose `height` disagrees is 0x13 before any hashing (V-N-16b).
 */
export function verifyInclusionForHeight(
  leaf: Digest,
  proof: InclusionProof,
  root: Digest,
  configuredHeight: number,
): void {
  if (proof.height !== configuredHeight) {
    throw new RegistryFailure("InclusionProofInvalid", `proof height ${proof.height} is not the log's ${configuredHeight}`);
  }
  verifyInclusion(leaf, proof, root);
}
